### ISSUE TITLE
Wrong unfulfilled lint expectation when using `#![expect(dead_code)]`

### ISSUE THREAD

**Issue by Alc-Alc** [rust-lang/rust#157093]
### Code

```Rust
#![expect(dead_code)]

struct SomeStruct;
trait SomeTrait {}
impl SomeTrait for SomeStruct {}


fn main() {
    
}
```

### Current output

```Shell
warning: this lint expectation is unfulfilled
 --> main.rs:1:11
  |
1 | #![expect(dead_code)]
  |           ^^^^^^^^^
  |
  = note: `#[warn(unfulfilled_lint_expectations)]` on by default

warning: 1 warning emitted
```

### Desired output

```Shell
No warning, as uncommenting this attribute will then raise 2 other warnings indicating that there some items that are unused.

Output without `#![expect(dead_code)]`


warning: struct `SomeStruct` is never constructed
 --> main.rs:3:8
  |
3 | struct SomeStruct;
  |        ^^^^^^^^^^
  |
  = note: `#[warn(dead_code)]` (part of `#[warn(unused)]`) on by default

warning: trait `SomeTrait` is never used
 --> main.rs:4:7
  |
4 | trait SomeTrait {}
  |       ^^^^^^^^^

warning: 2 warnings emitted
```

### Rationale and extra context

The way I understand, I `expect` there to be `dead_code` but the warning says otherwise, which if true, removing the `expect`ation should not raise any other `dead_code` warnings.

### Other cases

```Rust
Output changes depending on the presence of `#![expect(dead_code)]`
```

### Rust Version

```Shell
rustc 1.96.0 (ac68faa20 2026-05-25)
binary: rustc
commit-hash: ac68faa20c58cbccd01ee7208bf3b6e93a7d7f96
commit-date: 2026-05-25
host: aarch64-apple-darwin
release: 1.96.0
LLVM version: 22.1.2
```

### Anything else?

https://github.com/rust-lang/rust/issues/152370 was suggested by GitHub as a potential duplicate, but considering it is closed as completed and my occurrence is with the latest stable version I am considering this to be a different issue.

**Comment by Dnreikronos** [comment]
I reproduced this on stable (1.94.1 and 1.95.0) and confirmed it is already fixed on current nightly (1.97.0, 2026-04-23 and later).

Without `#![expect(dead_code)]`, both stable and nightly emit the two `dead_code` warnings, so the dead code itself is detected fine. With the attribute on stable, those warnings stop being attributed to the expectation, so `unfulfilled_lint_expectations` fires instead.

The trigger isn't the trait or the crate-level placement. The smallest reproduction is a dead type with any `impl` block, including an empty inherent one:

```rust
#![expect(dead_code)]
struct S;
impl S {}
fn main() {}
```

`struct S;` on its own (no impl) already reports the expectation as fulfilled, even on stable. Adding `impl S {}` is what breaks it.

The cause is in the dead-code pass. It runs a mark phase that builds the set of live symbols, then a warn phase that lints anything not in that set. An item carrying `allow`/`expect(dead_code)`, here inherited from the crate-level attribute, seeds the worklist as "comes from allow/expect". On stable, walking from such a seed could mark a referenced item live by mistake. The `impl` references `S` as its self type, so `S` lands in `live_symbols`, the warn phase skips it, no `dead_code` is emitted, and the expectation has nothing to fulfill.

This is the same root cause as #154324 and #152370, both closed by #154377 (merged 2026-04-21). I ran a batch of variants on a nightly built after that merge: crate-level and module-level expect, inherent impl, trait impl, struct fields plus methods, enums, generics, and a nested `pub` module. All of them report the expectation as fulfilled. Cases where nothing is actually dead (the type is used, or it is reachable) still report unfulfilled, so the fix is not over-suppressing.

So this looks like a duplicate of #154324. The fix is on master and nightly but has not reached stable yet, which is why it still shows up on 1.96.

**Comment by Alc-Alc** [comment]
@Dnreikronos thanks for the detailed response, are you suggesting I close the issue?

**Comment by Dnreikronos** [comment]
> [@Dnreikronos](https://github.com/Dnreikronos) thanks for the detailed response, are you suggesting I close the issue?

Yes, please, as it's already fixed on the latest nightly version.
